bioconda_utils/docker_utils.py

```diff
@@ -0,0 +1,402 @@
+                fout.write(open(pkg_resources.resource_filename(
+                    'bioconda_utils',
+                    'bioconda_utils-requirements.txt')
+                ).read())
+
+        with open(os.path.join(build_dir, "Dockerfile"), 'w') as fout:
+            fout.write(self.dockerfile_template.format(self=self))
+
+        logger.debug('Dockerfile:\n' + open(fout.name).read())
+
+        cmd = [
+            'docker', 'build',
+            '-t', self.tag,
+            build_dir
+        ]
+        p = sp.run(
+            cmd,
+            stdout=sp.PIPE,
+            stderr=sp.STDOUT,
+            check=True,
+            universal_newlines=True)
+
+        if p.returncode != 0:
+            raise ValueError(
+                p.stdout,
+                p.stderr)
+
+        logger.info('Built docker image tag=%s', self.tag)
+        shutil.rmtree(build_dir)
+        return p
+
+    def build_recipe(self, recipe_dir, build_args, env):
+        """
+        Build a single recipe.
+
+        Parameters
+        ----------
+
+        recipe_dir : str
+            Path to recipe that contains meta.yaml
+
+        build_args : str
+            Additional arguments to `conda build`. For example --channel,
+            --skip-existing, etc
+
+        env : dict
+            Environmental variables
+
+        Note that the binds are set up automatically to match the expectations
+        of the build script, and will use the currently-configured
+        self.container_staging and self.container_recipe.
+        """
+
+        # Attach the build args to self so that it can be filled in by the
+        # template.
+        if not isinstance(build_args, str):
+            raise ValueError('build_args must be str')
+        self.conda_build_args = build_args
+
+        # Write build script to tempfile
+        build_dir = tempfile.mkdtemp()
+        with open(os.path.join(build_dir, 'build_script.bash'), 'w') as fout:
+            fout.write(self.build_script_template.format(self=self))
+        build_script = fout.name
+        logger.debug('Container build script: \n%s', open(fout.name).read())
+
+
+        # Build the args for env vars. Note can also write these to tempfile
+        # and use --env-file arg, but using -e seems clearer in debug output.
+        env_list = []
+        for k, v in env.items():
+            env_list.append('-e')
+            env_list.append('{0}={1}'.format(k, v))
+
+        cmd = [
+            'docker', 'run',
+            '--net', 'host',
+            '-v', '{0}:/opt/build_script.bash'.format(build_script),
+            '-v', '{0}:{1}'.format(self.pkg_dir, self.container_staging),
+            '-v', '{0}:{1}'.format(recipe_dir, self.container_recipe),
+        ] + env_list + [
+            self.tag,
+            '/bin/bash', '/opt/build_script.bash',
+        ]
+
+        logger.debug('cmd: %s', cmd)
+        p = sp.run(cmd, stdout=sp.PIPE, stderr=sp.STDOUT, universal_newlines=True, check=True,)
```

Comment:
If `check=True`, `sp.run` will already check the exit code and fail with a CalledProcessError. Hence, I would think that DockerCalledProcessError will never be raised here. We can just remove it entirely.